packages/shared/src/deriveState.ts

```diff
@@ -0,0 +1,71 @@
+import type {
+  ActiveSessionResource,
+  InitialState,
+  OrganizationCustomPermissionKey,
+  OrganizationCustomRoleKey,
+  OrganizationResource,
+  Resources,
+  UserResource,
+} from '@clerk/types';
+
+export const deriveState = (clerkLoaded: boolean, state: Resources, initialState: InitialState | undefined) => {
+  if (!clerkLoaded && initialState) {
+    return deriveFromSsrInitialState(initialState);
+  }
+  return deriveFromClientSideState(state);
+};
+
+const deriveFromSsrInitialState = (initialState: InitialState) => {
+  const userId = initialState.userId;
+  const user = initialState.user as UserResource;
```

Comment:
Why do we need to typecast some properties in this function?

Reply:
Good question. Thanks for pointing this out! It was adapted from the React SDK, and I maintained the original structure (including typecasts).

My guess is we want to narrow the types if `InitialState` is provided? And also some of them are typed as strings (i.e. `orgRole`, `orgPermissions`) and typecasting it just for semantics / future-proofihng? Maybe @panteliselef 
 can help me answer this question 😄

Reply:
I asked because then I think `initialState` should be typed correctly in the first place 😊 

Maybe spend a little bit of time checking if we can fix this on a holistic level, otherwise the type-casts will be fine for now (if other things break)